### PR TITLE
mtda-service: move service code from mtda-cli to its own module

### PR DESCRIPTION
**Pull request by chombourger**
Closes: #116
Signed-off-by: Cedric Hombourger <cedric.hombourger@siemens.com>

### CODE REVIEW

**Comment by bhargavdas** [comment]
Did a quick test by rebasing these changes (and making changes to the yocto layers) with yocto qemuarm machine, everything seems to be working on a high level.
Except when the system starts the `mtda.service` is in failed state. I think this might due to missing config file.

**Comment by chombourger** [comment]
> Did a quick test by rebasing these changes (and making changes to the yocto layers) with yocto qemuarm machine, everything seems to be working on a high level. Except when the system starts the `mtda.service` is in failed state. I think this might due to missing config file.

Good to know. I assume we could add a conditional to the service to make it start only if /etc/mtda/config exists

File: mtda/main.py [hunk 1]
```diff
@@ -117,9 +117,15 @@ def __init__(self):
             self.config_files.append(os.path.join(home, '.mtda', 'config'))
 
         # Config file in /etc/mtda/config
-        if os.path.exists('/etc'):
+        if os.path.exists('/etc/mtda'):
             self.config_files.append(os.path.join('/etc', 'mtda', 'config'))
 
+        # Config fragments in /etc/mtda/config.d/
+        if os.path.exists('/etc/mtda/config.d'):
+            fragments = glob.glob('/etc/mtda/config.d/*.conf')
```
Comment (vj-kumar) [inline mtda/main.py:125]: What will be the parse order for multiple configs. I see that glob.glob does not return a sorted list. So the traditional 00_test.conf  01_test2.conf would not work?

Comment (chombourger) [reply mtda/main.py:125]: good point. added a call to `sorted()`

Comment (vj-kumar) [reply mtda/main.py:125]: I am not sure if we are already following some kind of precedence. From the document I understand that we parse in this order 
```
 * ``$HOME/.mtda/config``
 * ``/etc/mtda/config``
 * ``/etc/mtda/config.d/*.conf``
```

Does that mean '/etc/mtda/config.d/*.conf' has higher precedence? Or the one in HOME. Can we update document to explain the precedence?

**Comment by bhargavdas** [comment]
> > Did a quick test by rebasing these changes (and making changes to the yocto layers) with yocto qemuarm machine, everything seems to be working on a high level. Except when the system starts the `mtda.service` is in failed state. I think this might due to missing config file.
> 
> Good to know. I assume we could add a conditional to the service to make it start only if /etc/mtda/config exists

The condition to the service works, but still when the target (using yocto qemuarm here) boots/rebooted the `mtda.service` is in dead state initially and needs to started manually. Is this expected?
Here is the trimmed log,
1. target is booted first time without /etc/mtda/config
```
mtda login: root
root@mtda:~# systemctl list-units | grep mtda
  mtda-config.path                                                  loaded active     running         mtda-config.path
  mtda-config.service                                               loaded active     running         Service to monitor changes to mtda config file
  mtda.service                                                      loaded inactive   dead      start Multi-Tenant Device Access
root@mtda:~# systemctl start mtda
root@mtda:~# systemctl status mtda
* mtda.service - Multi-Tenant Device Access
     Loaded: loaded (/lib/systemd/system/mtda.service; enabled; vendor preset: enabled)
     Active: inactive (dead)
  Condition: start condition failed at Wed 2023-02-08 15:35:23 UTC; 24s ago
             `- ConditionPathExists=/etc/mtda/config was not met

Feb 08 15:35:23 mtda systemd[1]: Multi-Tenant Device Access was skipped because of a failed condition check (ConditionPathExists=/etc/mtda/config).
root@mtda:~# 
```
2. Created config
```
root@mtda:~#  cat /etc/mtda/config
[www]
port=9080
root@mtda:~# systemctl restart mtda
root@mtda:~# systemctl status mtda
* mtda.service - Multi-Tenant Device Access
     Loaded: loaded (/lib/systemd/system/mtda.service; enabled; vendor preset: enabled)
     Active: active (running) since Wed 2023-02-08 15:36:59 UTC; 1s ago
   Main PID: 257 (python3)
      Tasks: 1 (limit: 563)
     Memory: 5.9M
     CGroup: /system.slice/mtda.service
             `- 257 python3 /usr/sbin/mtda-service -n

Feb 08 15:36:59 mtda systemd[1]: Started Multi-Tenant Device Access.
root@mtda:~#
```
3. Rebooted the target
```
root@mtda:~# systemctl list-units | grep mtda
  mtda-config.path                                                  loaded active     running         mtda-config.path
  mtda-config.service                                               loaded active     running         Service to monitor changes to mtda config file
  mtda.service                                                      loaded inactive   dead      start Multi-Tenant Device Access
root@mtda:~# systemctl status mtda
* mtda.service - Multi-Tenant Device Access
     Loaded: loaded (/lib/systemd/system/mtda.service; enabled; vendor preset: enabled)
     Active: inactive (dead)
root@mtda:~#
root@mtda:~# cat /etc/mtda/config
[www]
port=9080
root@mtda:~#
```

**Comment by bhargavdas** [comment]
> The condition to the service works, but still when the target (using yocto qemuarm here) boots/rebooted the `mtda.service` is in dead state initially and needs to started manually. Is this expected? I doubt if this behavior is only observable in yocto builds, will trigger a nanopi build and check

Checked in nanopi, this is only a yocto issue maybe.

**Comment by chombourger** [comment]
> > The condition to the service works, but still when the target (using yocto qemuarm here) boots/rebooted the `mtda.service` is in dead state initially and needs to started manually. Is this expected? I doubt if this behavior is only observable in yocto builds, will trigger a nanopi build and check
> 
> Checked in nanopi, this is only a yocto issue maybe.

yes I believe our Isar images ship a mtda-config.service that cause systemd to restart mtda.service when /etc/mtda/config is changed

**Comment by bhargavdas** [comment]
> yes I believe our Isar images ship a mtda-config.service that cause systemd to restart mtda.service when /etc/mtda/config is changed

Even the yocto image ships this file and the service starts prior to this PR. I wonder if something might be missing in systemd files.

**Comment by bhargavdas** [comment]
> > yes I believe our Isar images ship a mtda-config.service that cause systemd to restart mtda.service when /etc/mtda/config is changed
> 
> Even the yocto image ships this file and the service starts prior to this PR. I wonder if something might be missing in systemd files.

Found it, the `Wants` section is blocking the start of the service.
```
Wants=network-online.target
```

**Comment by bhargavdas** [comment]
> > > yes I believe our Isar images ship a mtda-config.service that cause systemd to restart mtda.service when /etc/mtda/config is changed
> > 
> > 
> > Even the yocto image ships this file and the service starts prior to this PR. I wonder if something might be missing in systemd files.
> 
> Found it, the `Wants` section is blocking the start of the service.
> 
> ```
> Wants=network-online.target
> ```

The `Wants` unit requires the service to be active, but in yocto's case `network-online.target` dies after acknowledging that the network is online. In debian (I have checked in ubuntu) it remains alive always. Can this be changed to maybe, `network.target` ? Else I will have patch the file during the yocto build.

**Comment by bhargavdas** [comment]
> > > > yes I believe our Isar images ship a mtda-config.service that cause systemd to restart mtda.service when /etc/mtda/config is changed
> > > 
> > > 
> > > Even the yocto image ships this file and the service starts prior to this PR. I wonder if something might be missing in systemd files.
> > 
> > 
> > Found it, the `Wants` section is blocking the start of the service.
> > ```
> > Wants=network-online.target
> > ```
> 
> The `Wants` unit requires the service to be active, but in yocto's case (qemu) `network-online.target` dies after acknowledging that the network is online. In debian (I have checked in ubuntu) it remains alive always. Can this be changed to maybe, `network.target` ? Else I will have patch the file during the yocto build.

**Review by vj-kumar — APPROVED** [review]

**Pull request merged**